### PR TITLE
Decoder refactor

### PR DESCRIPTION
**Pull request by fintelia**
This does a bunch of refactoring with the hope of making it easier to change the public interface in the future. The user facing API should be unchanged, although more validation now happens in Decoder::new and Decoder::next_image, so some errors may be caught earlier. (This in particular broke a few fuzzer generated test cases: the images are now rejected as corrupt long before execution reaches the code those were meant to test.)

At a high level, many of the fields of Decoder are now moved into a new Image struct with represents a single subimage of a tiff file. It is constructed with the `Image::from_reader` method. Right now the code still lazily loads subimages, but most of the code is in place to do things eagerly instead.

### CODE REVIEW

**Review by HeroicKatora — COMMENTED** [review]
Could you explain a bit more detailed how you intend to leverage the new structure, and in particular `Image`? It seems we're doing _more_ things eagerly already. This isn't self-evidently a good choice, for performance reasons and due to the observed behavior changes. Could you also motivate the advantages of an eager approach?

File: tests/decode_images.rs [hunk 1]
```diff
@@ -343,26 +343,6 @@ fn test_bad_strip_count() {
...
 }
 
-#[test]
-fn test_too_big_buffer_size() {
-    let image = [
-        73, 73, 42, 0, 8, 0, 0, 0, 8, 0, 0, 1, 4, 0, 1, 0, 0, 0, 99, 255, 255, 254, 1, 1, 4, 0, 1,
-        0, 0, 0, 158, 0, 0, 251, 3, 1, 3, 255, 254, 255, 255, 0, 1, 0, 0, 0, 6, 1, 3, 0, 1, 0, 0,
-        0, 0, 0, 0, 0, 17, 1, 4, 0, 9, 0, 0, 0, 0, 0, 0, 0, 2, 1, 3, 0, 2, 0, 0, 0, 63, 0, 0, 0,
-        22, 1, 4, 0, 1, 0, 0, 0, 44, 0, 0, 0, 23, 1, 4, 0, 0, 0, 0, 0, 0, 0, 2, 1, 3, 1, 0, 178,
-        178,
```
[the remark this answers is not in the document]
Comment (fintelia) [reply tests/decode_images.rs:353]: The problem is that this an the other removed inputs were generated by a fuzzer, and are actually corrupt for a bunch of reasons. So for instance, in this case it was only by chance that our code produced a LimitsExceeded error. If you validate the file with `tiffinfo`, you actually get a completely different error:

>  TIFFReadDirectoryCheckOrder: Warning, Invalid TIFF directory; tags are not sorted in ascending order.
TIFFReadDirectory: Incompatible type for "Compression".

[the remark this answers is not in the document]
Comment (HeroicKatora) [reply tests/decode_images.rs:353]: Aye, but they were generated by fuzzer including for the reason that they had crashed, or oom'ed, the decoder at some point. I'd like to preserve them as test cases for those checks as much as possible. Feel free to drop specific error checks, though, sure.

**Comment by fintelia** [comment]
> Could you explain a bit more detailed how you intend to leverage the new structure, and in particular Image?

By pulling out all the information about a subimage into the `Image` struct, we can represent all the subimages in a TIFF file as a `Vec<Image>`. There's a couple directions we could go with that. We could add a Decoder::prev_image method that seeks backwards, we eagerly load all the subimages in _Decoder::new_, we could be extra lazy and not even load the first subimage until needed.

Another direction is parallel decoding of multiple chunks or images. This restructuring ensures that all access to the subimage metadata happens through a shared reference, which is somewhat of a prerequisite for that. However at this point, I haven't worked out what the full API for parallel decoding might look like. 

> It seems we're doing more things eagerly already. This isn't self-evidently a good choice, for performance reasons and due to the observed behavior changes. Could you also motivate the advantages of an eager approach?

I'm open to thinking this through this in more detail. Performance-wise I don't think it will usually make much of a difference for most use cases because reading/decompressing the image data is the biggest cost. The behavior changes are mostly noticing errors in a different order, but this change does also preclude manually reading tags of corrupted files.

**Review by HeroicKatora — APPROVED** [review]
> We could add a Decoder::prev_image method that seeks backwards, we eagerly load all the subimages in Decoder::new, we could be extra lazy and not even load the first subimage until needed.

That sounds pretty good. I can see how those two could work, and be useful.

**Pull request merged**